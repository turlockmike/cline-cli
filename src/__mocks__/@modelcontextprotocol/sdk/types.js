const ErrorCode = {
  InvalidRequest: 'INVALID_REQUEST',
  MethodNotFound: 'METHOD_NOT_FOUND',
  ExecutionError: 'EXECUTION_ERROR',
  InvalidParams: 'INVALID_PARAMS'
};

class McpError extends Error {
  constructor(code, message) {
    super(message);
    this.code = code;
    this.name = 'McpError';
  }
}

const ListToolsRequestSchema = {
  name: 'listTools',
  method: 'tools/list',
  params: {}
};

const CallToolRequestSchema = {
  name: 'callTool',
  method: 'tools/call',
  params: {
    name: '',
    arguments: {}
  }
};

module.exports = {
  ErrorCode,
  McpError,
  ListToolsRequestSchema,
  CallToolRequestSchema
};