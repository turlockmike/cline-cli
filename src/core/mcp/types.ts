import type { ToolExecutionOptions } from 'ai';

/**
 * Base interface for MCP tool responses
 */
export interface McpToolResponse<T = any> {
    content: Array<{
        type: string;
        text: string;
    }>;
    isError: boolean;
}

/**
 * Parsed MCP tool response
 */
export interface ParsedMcpToolResponse<T = any> {
    data: T;
    raw: McpToolResponse;
}

/**
 * Base tool interface matching the AI package's Tool type
 */
export interface BaseTool<TArgs = any, TResult = any> {
    description?: string;
    parameters: Record<string, any>;
    execute: (args: TArgs, options: ToolExecutionOptions) => Promise<TResult>;
}

/**
 * MCP tool with proper typing
 */
export interface McpTool<TArgs = any, TResult = any> extends BaseTool<TArgs, ParsedMcpToolResponse<TResult>> {
    description: string;
    parameters: Record<string, any>;
}

/**
 * Jira ticket interface
 */
export interface JiraTicket {
    key: string;
    summary: string;
    status: string;
    type: string;
    priority: string;
    assignee?: {
        name: string;
        email: string;
    };
    reporter?: {
        name: string;
        email: string;
    };
    description?: string;
    created?: string;
    updated?: string;
    url?: string;
}

/**
 * Arguments for Jira get ticket tool
 */
export interface JiraGetTicketArgs {
    ticketId: string;
}

/**
 * Jira get ticket tool type
 */
export type JiraGetTicketTool = McpTool<JiraGetTicketArgs, JiraTicket>;