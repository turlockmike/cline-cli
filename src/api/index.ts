import { Anthropic } from "@anthropic-ai/sdk"
import { GlamaProvider } from "./providers/glama"
import { ModelConfiguration as ModelConfiguration, ModelInfo } from "../shared/api"
import { AnthropicProvider } from "./providers/anthropic"
import { AwsBedrockProvider } from "./providers/bedrock"
import { OpenRouterProvider } from "./providers/openrouter"
import { VertexProvider } from "./providers/vertex"
import { OpenAiProvider } from "./providers/openai"
import { OllamaProvider } from "./providers/ollama"
import { LmStudioProvider } from "./providers/lmstudio"
import { GeminiProvider } from "./providers/gemini"
import { OpenAiNativeProvider } from "./providers/openai-native"
import { DeepSeekProvider } from "./providers/deepseek"
import { MistralProvider } from "./providers/mistral"
import { ApiStream } from "./transform/stream"

export interface SingleCompletionHandler {
	completePrompt(prompt: string): Promise<string>
}

export interface ModelProvider {
	createMessage(systemPrompt: string, messages: Anthropic.Messages.MessageParam[]): ApiStream
	getModel(): { id: string; info: ModelInfo }
}

export function modelProviderFromConfig(configuration: ModelConfiguration): ModelProvider {
	const { apiProvider, ...options } = configuration
	switch (apiProvider) {
		case "anthropic":
			return new AnthropicProvider(options)
		case "glama":
			return new GlamaProvider(options)
		case "openrouter":
			return new OpenRouterProvider(options)
		case "bedrock":
			return new AwsBedrockProvider(options)
		case "vertex":
			return new VertexProvider(options)
		case "openai":
			return new OpenAiProvider(options)
		case "ollama":
			return new OllamaProvider(options)
		case "lmstudio":
			return new LmStudioProvider(options)
		case "gemini":
			return new GeminiProvider(options)
		case "openai-native":
			return new OpenAiNativeProvider(options)
		case "deepseek":
			return new DeepSeekProvider(options)
		default:
			throw new Error(`Unsupported API provider: ${apiProvider}`)
	}
}
