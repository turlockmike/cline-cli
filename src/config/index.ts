// Configuration Paths
export { getConfigPaths, createConfigDirectories } from './configPaths';

// Profile Management
export { ProfileManager } from './ProfileManager';
export { Profile, ProfileSchema, ProfilesConfig, ProfilesConfigSchema, ProfileOptions, ProfileOptionsSchema } from './profileConfig';

// Tool Management
export { ToolManager } from './ToolManager';
export { ToolSetConfig, ToolsConfig, ToolSetConfigSchema, ToolsConfigSchema } from './toolConfig';

// Agent Management
export { AgentManager } from './AgentManager';
export { AgentConfig, ModelConfig, ModelParameters, AgentConfigSchema, ModelConfigSchema } from './agentConfig';

// Task Management
export { TaskManager } from './TaskManager';
export { TaskConfig, TaskTemplate, TaskConfigSchema, TaskTemplateSchema } from './taskConfig';

// Configuration Loader
export { ConfigLoader, CliOptions } from './ConfigLoader';

// First Run Manager
export { FirstRunManager } from './FirstRunManager';