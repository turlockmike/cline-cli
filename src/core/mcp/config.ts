import { z } from 'zod';

/**
 * Configuration for an MCP server
 */
export interface McpServerConfig {
  /** Command to run the server */
  command: string;
  /** Arguments to pass to the command */
  args: string[];
  /** Environment variables to set when running the server */
  env?: Record<string, string>;
  /** List of tool names to disable from this server */
  disabledTools?: string[];
}

/**
 * Complete MCP configuration
 */
export interface McpConfig {
  /** Map of server names to their configurations */
  mcpServers: Record<string, McpServerConfig>;
}

/**
 * Zod schema for validating MCP server configuration
 */
export const McpServerConfigSchema = z.object({
  command: z.string().min(1),
  args: z.array(z.string()),
  env: z.record(z.string()).optional(),
  disabledTools: z.array(z.string()).optional(),
});

/**
 * Zod schema for validating complete MCP configuration
 */
export const McpConfigSchema = z.object({
  mcpServers: z.record(McpServerConfigSchema),
});

/**
 * Type guard to check if a value is a valid MCP configuration
 */
export function isMcpConfig(value: unknown): value is McpConfig {
  return McpConfigSchema.safeParse(value).success;
} 