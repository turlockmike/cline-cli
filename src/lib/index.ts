export * from './types';
export * from './tools-deprecated';
export * from './parser';

// Re-export specific types and implementations
export type { ParsedTool } from './parser';
export { MessageParser } from './parser';
export { AVAILABLE_TOOLS } from './tools-deprecated';