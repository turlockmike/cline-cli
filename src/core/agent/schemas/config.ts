import { z } from 'zod';
import { ModelProvider } from '../../../api';
import { ModelConfiguration } from '../../../shared/api';

// Schema for ModelConfiguration
const modelConfigurationSchema = z.object({
  apiProvider: z.string(),
  apiModelId: z.string().optional(),
  id: z.string().optional(),
}).passthrough(); // Allow additional properties from ModelProviderOptions

// Schema for ModelProvider interface
const modelProviderSchema = z.object({
  createMessage: z.function(),
  getModel: z.function(),
});

// Combined model schema that accepts either ModelProvider or ModelConfiguration
const modelSchema = z.union([modelProviderSchema, modelConfigurationSchema]);

export const streamingConfigSchema = z.object({
  enabled: z.boolean(),
  chunkSize: z.number().positive().optional(),
  maxDelay: z.number().positive().optional(),
});

// Schema for SystemPromptSection
const systemPromptSectionSchema = z.object({
  name: z.string(),
  content: z.string(),
  order: z.number(),
  enabled: z.boolean(),
});

// Schema for SystemPromptConfig sections
const systemPromptSectionsSchema = z.object({
  capabilities: z.object({
    computerUse: z.boolean().optional(),
    mcpSupport: z.boolean().optional(),
    diffStrategy: z.string().optional(), // Changed from instanceof to string
  }).optional(),
  customInstructions: z.object({
    language: z.string().optional(),
    instructions: z.string().optional(),
  }).optional(),
  rules: z.object({
    additionalRules: z.array(z.string()).optional(),
    disabledRules: z.array(z.string()).optional(),
  }).optional(),
  systemInfo: z.object({
    additionalInfo: z.record(z.string()).optional(),
  }).optional(),
  toolUse: z.object({
    additionalTools: z.array(z.string()).optional(),
    disabledTools: z.array(z.string()).optional(),
  }).optional(),
  mcpServers: z.object({
    servers: z.any().optional(), // McpHub instance
  }).optional(),
  objective: z.object({
    customObjective: z.string().optional(),
  }).optional(),
}).optional();

// Schema for SystemPromptConfig
const systemPromptConfigSchema = z.object({
  sections: systemPromptSectionsSchema,
  customSections: z.array(systemPromptSectionSchema).optional(),
  options: z.object({
    sectionOrder: z.array(z.string()).optional(),
    disabledSections: z.array(z.string()).optional(),
  }).optional(),
}).optional();

// Updated agent config schema to use UnifiedTool and include systemPrompt
export const agentConfigSchema = z.object({
  name: z.string(),
  model: z.union([
    z.custom<ModelProvider>((val) => val && typeof (val as any).createMessage === 'function'),
    z.object({
      apiProvider: z.string(),
      apiModelId: z.string()
    })
  ]),
  tools: z.array(z.any()).optional(),
  role: z.string().optional(),
  customInstructions: z.string().optional(),
  systemPromptConfig: z.any().optional(),
  streaming: streamingConfigSchema.optional(),
  maxRetries: z.number().int().nonnegative().optional(),
  timeout: z.number().positive().optional(),
}).strict();