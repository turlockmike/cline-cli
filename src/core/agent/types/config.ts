import { z } from 'zod';
import { UnifiedTool } from '../../../lib/types';
import { ModelConfiguration, ModelProvider as ApiModelProvider } from '../../../shared/api';
import { ModelProvider } from '../../../api';
import { SystemPromptConfig } from '../../prompts/prompt-builder';
import { Thread } from '../../thread/thread';

/**
 * Model configuration for the agent - can be either a ModelProvider instance or a ModelConfiguration
 */
export type ModelConfig = ModelProvider | ModelConfiguration;

/**
 * Configuration options for response streaming
 */
export interface StreamingConfig {
  /** Whether to enable streaming responses */
  enabled: boolean;
  /** Chunk size for streaming responses in bytes */
  chunkSize?: number;
  /** Maximum time to wait between chunks in milliseconds */
  maxDelay?: number;
}

/**
 * Main configuration interface for the Agent
 */
export interface AgentConfig {
  /** The name of the agent */
  name: string;
  /** Model configuration - can be either a ModelProvider instance or a ModelConfiguration */
  model: ModelConfig;
  /** List of tools to use - must be UnifiedTool instances */
  tools?: UnifiedTool[];
  /** Streaming configuration */
  streaming?: StreamingConfig;
  /** Maximum number of retries for failed operations */
  maxRetries?: number;
  /** Timeout for operations in milliseconds */
  timeout?: number;
  /** The role of the agent to be used in the system prompt */
  role?: string;
  /** The custom instructions for the agent to be used in the system prompt */
  customInstructions?: string;
  /** The system prompt configuration. By default, the system prompt will be generated using the role and custom instructions and a minimal set of rules for tool use. */
  systemPromptConfig?: SystemPromptConfig;
}

/**
 * Role type for task messages
 */
export type TaskRole = 'user' | 'assistant' | 'system';

/**
 * File context for tasks
 */
export interface FileContext {
  name: string;
  content: Buffer;
  type: string;
}

/**
 * Task input configuration
 */
export interface TaskInput<TOutput = string> {
  /** The role of the message sender */
  role: TaskRole;
  /** The content of the task */
  content: string;
  /** The thread to use for context */
  thread?: Thread;
  /** Whether to stream the response */
  stream?: boolean;
  /** Additional context for the task */
  context?: Record<string, unknown> | FileContext[];
  /** Schema for validating the output */
  outputSchema?: z.ZodSchema<TOutput>;
}
