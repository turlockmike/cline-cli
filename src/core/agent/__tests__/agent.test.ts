import { Agent } from ".."
import { AgentConfig, TaskInput, Thread } from "../types/config"
import { z } from "zod"
import { MockProvider, MockTool } from "../../../lib/testing"
import { ModelProvider } from "../../../api"

describe("Agent", () => {
	let mockProvider: MockProvider
	let mockTool: MockTool
	let mockToolWithInit: MockTool
	let validConfigWithProvider: AgentConfig
	let validConfigWithModelConfig: AgentConfig

	beforeEach(() => {
		mockProvider = new MockProvider()
		mockTool = MockTool.createBasic("mock_tool")
		mockToolWithInit = MockTool.createBasic("mock_tool_init")
		mockToolWithInit.setInitialize(async () => {
			/* mock initialization */
		})

		validConfigWithProvider = {
			name: "test-agent",
			model: mockProvider as ModelProvider,
			tools: [mockTool, mockToolWithInit],
			role: "You are Hataraku, a highly skilled software engineer with extensive knowledge in many programming languages, frameworks, design patterns, and best practices.",
			customInstructions: "You should always speak and think in the English language.",
		}

		validConfigWithModelConfig = {
			name: "test-agent",
			model: {
				apiProvider: "anthropic",
				apiModelId: "claude-3-5-sonnet-20241022",
			},
			tools: [mockTool, mockToolWithInit],
			role: "You are Hataraku, a highly skilled software engineer with extensive knowledge in many programming languages, frameworks, design patterns, and best practices.",
			customInstructions: "You should always speak and think in the English language.",
		}
	})

	const validTaskInput: TaskInput & { stream?: false } = {
		role: "user",
		content: "test task",
		stream: false,
	}

	describe("constructor", () => {
		it("should create an instance with ModelProvider", () => {
			const agent = new Agent(validConfigWithProvider)
			expect(agent).toBeInstanceOf(Agent)
			expect(agent.getConfig()).toEqual(validConfigWithProvider)
			expect(agent.getModelProvider()).toBe(mockProvider)
		})

		it("should create an instance with ModelConfiguration", () => {
			const agent = new Agent(validConfigWithModelConfig)
			expect(agent).toBeInstanceOf(Agent)
			expect(agent.getConfig()).toEqual(validConfigWithModelConfig)
			expect(agent.getModelProvider()).toBeDefined()
		})

		it("should throw error with invalid config", () => {
			const invalidConfig = {
				model: {
					apiProvider: "invalid-provider",
				},
			}

			expect(() => new Agent(invalidConfig as any)).toThrow("Invalid agent configuration")
		})
	})

	describe("initialize", () => {
		it("should initialize successfully with ModelProvider", async () => {
			const agent = new Agent(validConfigWithProvider)
			await agent.initialize()
			expect(agent.getLoadedTools()).toContain("mock_tool")
			expect(agent.getLoadedTools()).toContain("mock_tool_init")
			expect(mockToolWithInit.getInitializeCallCount()).toBe(1)
		})

		it("should initialize successfully with ModelConfiguration", async () => {
			const agent = new Agent(validConfigWithModelConfig)
			await agent.initialize()
			expect(agent.getLoadedTools()).toContain("mock_tool")
			expect(agent.getLoadedTools()).toContain("mock_tool_init")
			expect(mockToolWithInit.getInitializeCallCount()).toBe(1)
		})

		it("should only initialize once", async () => {
			const agent = new Agent(validConfigWithProvider)
			await agent.initialize()
			await agent.initialize()
			expect(agent.getLoadedTools()).toContain("mock_tool")
			expect(mockToolWithInit.getInitializeCallCount()).toBe(1)
		})
	})

	describe("task", () => {
		it("should include role and custom instructions in system prompt", async () => {
			mockProvider.clearResponses().mockResponse("test response")
			const agent = new Agent(validConfigWithProvider)
			await agent.initialize()

			await agent.task(validTaskInput)
			const call = mockProvider.getCall(0)!
			expect(call).toBeTruthy()
			console.log("call", call)

			// Verify system prompt content
			expect(call.systemPrompt).toMatchSnapshot()
		})

		it("should execute task successfully", async () => {
			mockProvider.clearResponses().mockResponse("test response")
			mockTool.mockResponse({ result: "tool success" })

			const agent = new Agent(validConfigWithProvider)
			await agent.initialize()

			const result = await agent.task(validTaskInput)
			expect(result).toBe("test response")
			expect(mockProvider.getCallCount()).toBe(1)

			const call = mockProvider.getCall(0)!
			expect(call.systemPrompt).not.toContain("test task")
			expect(call.messages).toHaveLength(1)
			expect(call.messages[0]).toEqual({
				role: "user",
				content: "<task>test task</task>",
			})
		})

		it("should handle streaming task", async () => {
			mockProvider.clearResponses().mockResponse("test response")
			const agent = new Agent(validConfigWithProvider)
			await agent.initialize()

			const streamingInput: TaskInput & { stream: true } = {
				role: "user",
				content: "test task",
				stream: true,
			}

			const result = await agent.task(streamingInput)
			expect(result).toBeDefined()
			expect(result[Symbol.asyncIterator]).toBeDefined()
		})

		it("should handle task with thread context", async () => {
			mockProvider.clearResponses().mockResponse("test response")
			const agent = new Agent(validConfigWithProvider)
			const thread = new Thread()
			thread.addContext({ key: "test", content: { value: "test" } })

			await agent.initialize()
			const result = await agent.task({ ...validTaskInput, thread })
			expect(result).toBe("test response")

			const call = mockProvider.getCall(0)!
			expect(call.messages).toHaveLength(2) // Context message + task message
			expect(call.messages[0].content).toContain("test") // Context included
		})

		it("should handle task with output schema", async () => {
			mockProvider.clearResponses().mockResponse('{"result": "test"}')
			const agent = new Agent(validConfigWithProvider)
			const outputSchema = z.object({
				result: z.string(),
			})

			await agent.initialize()
			const result = await agent.task({ ...validTaskInput, outputSchema })
			expect(result).toEqual({ result: "test" })
		})

		it("should throw error for invalid output schema", async () => {
			mockProvider.clearResponses().mockResponse("invalid json")
			const agent = new Agent(validConfigWithProvider)
			const outputSchema = z.object({
				result: z.string(),
			})

			await agent.initialize()
			await expect(agent.task({ ...validTaskInput, outputSchema })).rejects.toThrow(
				"Failed to parse response with schema"
			)
		})

		it("should handle model errors", async () => {
			mockProvider.clearResponses().mockError("Model error")
			const agent = new Agent(validConfigWithProvider)
			await agent.initialize()

			await expect(agent.task(validTaskInput)).rejects.toThrow("Model error")
		})

		it("should handle tool errors", async () => {
			mockProvider.clearResponses().mockResponse("test response")
			mockTool.mockError(new Error("Tool error"))

			const agent = new Agent(validConfigWithProvider)
			await agent.initialize()

			const result = await agent.task(validTaskInput)
			expect(result).toBe("test response")
			expect(mockTool.getCallCount()).toBe(0)
		})

		it("should include schema validation instructions in system prompt when output schema is provided", async () => {
			mockProvider.clearResponses().mockResponse('{"result": "test"}')
			const agent = new Agent(validConfigWithProvider)
			const outputSchema = z.object({
				result: z.string(),
			})

			await agent.initialize()
			const result = await agent.task({ ...validTaskInput, outputSchema })
			expect(result).toEqual({ result: "test" })
			
			const call = mockProvider.getCall(0)!
      console.log('last Message', call.messages[call.messages.length - 1])
			expect(call.systemPrompt).toContain("Schema Validation and Output Formatting")
			expect(call.systemPrompt).toContain("Your response must be valid JSON that matches the schema exactly")
			expect(call.systemPrompt).toContain("Do not include any additional text, explanations, or formatting around the JSON")
			expect(call.systemPrompt).toContain("Ensure all required fields specified in the schema are present")
			expect(call.systemPrompt).toContain("Only include fields that are defined in the schema")
			expect(call.systemPrompt).toContain("Use the correct data types for each field as specified in the schema")
			expect(call.systemPrompt).toContain("For streaming responses, each chunk must be valid JSON that matches the schema")
      expect(call.systemPrompt).toContain('You will be given a task in a <task></task> tag. You will also be optionally given an output schema in a <output_schema></output_schema> tag.')
      expect(call.systemPrompt).toContain("When calling attempt_completion, ensure the result is valid JSON that matches the schema")
      expect(call.messages[0].content).toEqual('<task>test task</task><output_schema>{"result": z.string()}</output_schema>')
		})

    it('with a slightly more complex schema', async () => {
      mockProvider.clearResponses().mockResponse('{"foo": 123}')
      const agent = new Agent(validConfigWithProvider)
      const outputSchema = z.object({
        foo: z.number(),
      })

      await agent.initialize()
      const result = await agent.task({ ...validTaskInput, outputSchema })
      expect(result).toEqual({ foo: 123 })

      const call = mockProvider.getCall(0)!
      expect(call.messages[0].content).toEqual('<task>test task</task><output_schema>{"foo": z.number()}</output_schema>')
    })
	})
})
