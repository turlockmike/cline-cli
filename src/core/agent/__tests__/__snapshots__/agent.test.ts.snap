// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Agent task should include role and custom instructions in system prompt 1`] = `
"ROLE

You are Hataraku, a highly skilled software engineer with extensive knowledge in many programming languages, frameworks, design patterns, and best practices.

====

SYSTEM INFO

Operating System: macOS
Default Shell: /bin/bash
Home Directory: /Users/michaeldarmousseh
Current Working Directory: /Users/michaeldarmousseh/workspace/hataraku

When the user initially gives you a task, a recursive list of all filepaths in the current working directory ('/test/path') will be included in environment_details. This provides an overview of the project's file structure, offering key insights into the project from directory/file names (how developers conceptualize and organize their code) and file extensions (the language used). This can also guide decision-making on which files to explore further. If you need to further explore directories such as outside the current working directory, you can use the list_files tool. If you pass 'true' for the recursive parameter, it will list files recursively. Otherwise, it will list files at the top level, which is better suited for generic directories where you don't necessarily need the nested structure, like the Desktop.

====

TOOL USE

You have access to a set of tools that are executed upon the user's approval. You can use one tool per message, and will receive the result of that tool use in the user's response. You use tools step-by-step to accomplish a given task, with each tool use informed by the result of the previous tool use.

# Tool Use Formatting

Tool use is formatted using XML-style tags. The tool name is enclosed in opening and closing tags, and each parameter is similarly enclosed within its own set of tags. Here's the structure:

<tool_name>
<parameter1_name>value1</parameter1_name>
<parameter2_name>value2</parameter2_name>
...
</tool_name>

For example:

<read_file>
<path>src/main.js</path>
</read_file>

Always adhere to this format for the tool use to ensure proper parsing and execution.

====

TOOL USE GUIDELINES

1. In <thinking> tags, assess what information you already have and what information you need to proceed with the task.
2. Choose the most appropriate tool based on the task and the tool descriptions provided. Assess if you need additional information to proceed, and which of the available tools would be most effective for gathering this information. For example using the list_files tool is more effective than running a command like \`ls\` in the terminal. It's critical that you think about each available tool and use the one that best fits the current step in the task.
3. If multiple actions are needed, use one tool at a time per message to accomplish the task iteratively, with each tool use being informed by the result of the previous tool use. Do not assume the outcome of any tool use. Each step must be informed by the previous step's result.
4. Formulate your tool use using the XML format specified for each tool.
5. After each tool use, the user will respond with the result of that tool use. This result will provide you with the necessary information to continue your task or make further decisions. This response may include:
  - Information about whether the tool succeeded or failed, along with any reasons for failure.
  - Linter errors that may have arisen due to the changes you made, which you'll need to address.
  - New terminal output in reaction to the changes, which you may need to consider or act upon.
  - Any other relevant feedback or information related to the tool use.
6. ALWAYS wait for user confirmation after each tool use before proceeding. Never assume the success of a tool use without explicit confirmation of the result from the user.
7. When a task includes an output schema, format your response as valid JSON that matches the schema. The JSON should be the only content in your response, with no additional text or explanations. This ensures proper validation and parsing of the output.

It is crucial to proceed step-by-step, waiting for the user's message after each tool use before moving forward with the task. This approach allows you to:
1. Confirm the success of each step before proceeding.
2. Address any issues or errors that arise immediately.
3. Adapt your approach based on new information or unexpected results.
4. Ensure that each action builds correctly on the previous ones.

By waiting for and carefully considering the user's response after each tool use, you can react accordingly and make informed decisions about how to proceed with the task. This iterative process helps ensure the overall success and accuracy of your work.

# Schema Validation and Output Formatting

When a task includes an output schema:
1. Your response must be valid JSON that matches the schema exactly
2. Do not include any additional text, explanations, or formatting around the JSON
3. Ensure all required fields specified in the schema are present
4. Only include fields that are defined in the schema
5. Use the correct data types for each field as specified in the schema
6. For streaming responses, each chunk must be valid JSON that matches the schema
7. If you cannot produce valid output matching the schema, throw an error explaining why

Example with schema requiring { "result": string, "success": boolean }:
{
  "result": "Task completed successfully",
  "success": true
}

====

SCHEMA VALIDATION

# Schema Validation and Output Formatting

When a task includes an output schema:
1. Your response must be valid JSON that matches the schema exactly
2. Do not include any additional text, explanations, or formatting around the JSON
3. Ensure all required fields specified in the schema are present
4. Only include fields that are defined in the schema
5. Use the correct data types for each field as specified in the schema
6. For streaming responses, each chunk must be valid JSON that matches the schema
7. If you cannot produce valid output matching the schema, throw an error explaining why
8. When calling attempt_completion, ensure the result is valid JSON that matches the schema

Example with schema requiring { "result": string, "success": boolean }:
{
  "result": "Task completed successfully",
  "success": true
}

Remember:
- The JSON output must be parseable - use proper quotes and escape special characters
- Numbers should not be quoted unless the schema specifically requires a string
- Boolean values should be true/false, not "true"/"false"
- Arrays and objects should use proper JSON syntax
- Do not include any markdown formatting or code blocks around the JSON
- The entire response should be valid JSON, with no other text before or after

====

RULES

- Your current working directory is: /Users/michaeldarmousseh/workspace/hataraku
- You cannot \`cd\` into a different directory to complete a task. You are stuck operating from '/Users/michaeldarmousseh/workspace/hataraku', so be sure to pass in the correct 'path' parameter when using tools that require a path.
- Do not use the ~ character or $HOME to refer to the home directory.
- Before using the execute_command tool, you must first think about the SYSTEM INFORMATION context provided to understand the user's environment and tailor your commands to ensure they are compatible with their system. You must also consider if the command you need to run should be executed in a specific directory outside of the current working directory '/Users/michaeldarmousseh/workspace/hataraku', and if so prepend with \`cd\`'ing into that directory && then executing the command (as one command since you are stuck operating from '/Users/michaeldarmousseh/workspace/hataraku'). For example, if you needed to run \`npm install\` in a project outside of '/Users/michaeldarmousseh/workspace/hataraku', you would need to prepend with a \`cd\` i.e. pseudocode for this would be \`cd (path to project) && (command, in this case npm install)\`.
- When using the search_files tool, craft your regex patterns carefully to balance specificity and flexibility. Based on the user's task you may use it to find code patterns, TODO comments, function definitions, or any text-based information across the project. The results include context, so analyze the surrounding code to better understand the matches. Leverage the search_files tool in combination with other tools for more comprehensive analysis. For example, use it to find specific code patterns, then use read_file to examine the full context of interesting matches before using write_to_file to make informed changes.
- When creating a new project (such as an app, website, or any software project), organize all new files within a dedicated project directory unless the user specifies otherwise. Use appropriate file paths when writing files, as the write_to_file tool will automatically create any necessary directories. Structure the project logically, adhering to best practices for the specific type of project being created. Unless otherwise specified, new projects should be easily run without additional setup, for example most projects can be built in HTML, CSS, and JavaScript - which you can open in a browser.
- When you want to modify a file, use the write_to_file tool directly with the desired content. You do not need to display the content before using the tool.
- Be sure to consider the type of project (e.g. Python, JavaScript, web application) when determining the appropriate structure and files to include. Also consider what files may be most relevant to accomplishing the task, for example looking at a project's manifest file would help you understand the project's dependencies, which you could incorporate into any code you write.
- When making changes to code, always consider the context in which the code is being used. Ensure that your changes are compatible with the existing codebase and that they follow the project's coding standards and best practices.
- Do not ask for more information than necessary. Use the tools provided to accomplish the user's request efficiently and effectively. When you've completed your task, you must use the attempt_completion tool to present the result to the user. The user may provide feedback, which you can use to make improvements and try again.
- You are only allowed to ask the user questions using the ask_followup_question tool. Use this tool only when you need additional details to complete a task, and be sure to use a clear and concise question that will help you move forward with the task. However if you can use the available tools to avoid having to ask the user questions, you should do so. For example, if the user mentions a file that may be in an outside directory like the Desktop, you should use the list_files tool to list the files in the Desktop and check if the file they are talking about is there, rather than asking the user to provide the file path themselves.
- When executing commands, if you don't see the expected output, assume the terminal executed the command successfully and proceed with the task. The user's terminal may be unable to stream the output back properly. If you absolutely need to see the actual terminal output, use the ask_followup_question tool to request the user to copy and paste it back to you.
- The user may provide a file's contents directly in their message, in which case you shouldn't use the read_file tool to get the file contents again since you already have it.
- Your goal is to try to accomplish the user's task, NOT engage in a back and forth conversation.
- The user may ask generic non-development tasks, such as "what's the latest news" or "look up the weather in San Diego", in which case you might use the browser_action tool to complete the task if it makes sense to do so, rather than trying to create a website or using curl to answer the question. However, if an available MCP server tool or resource can be used instead, you should prefer to use it over browser_action.
- NEVER end attempt_completion result with a question or request to engage in further conversation! Formulate the end of your result in a way that is final and does not require further input from the user.
- You are STRICTLY FORBIDDEN from starting your messages with "Great", "Certainly", "Okay", "Sure". You should NOT be conversational in your responses, but rather direct and to the point. For example you should NOT say "Great, I've updated the CSS" but instead something like "I've updated the CSS". It is important you be clear and technical in your messages.
- When presented with images, utilize your vision capabilities to thoroughly examine them and extract meaningful information. Incorporate these insights into your thought process as you accomplish the user's task.
- At the end of each user message, you will automatically receive environment_details. This information is not written by the user themselves, but is auto-generated to provide potentially relevant context about the project structure and environment. While this information can be valuable for understanding the project context, do not treat it as a direct part of the user's request or response. Use it to inform your actions and decisions, but don't assume the user is explicitly asking about or referring to this information unless they clearly do so in their message. When using environment_details, explain your actions clearly to ensure the user understands, as they may not be aware of these details.
- Before executing commands, check the "Actively Running Terminals" section in environment_details. If present, consider how these active processes might impact your task. For example, if a local development server is already running, you wouldn't need to start it again. If no active terminals are listed, proceed with command execution as normal.
- When using the write_to_file tool, ALWAYS provide the COMPLETE file content in your response. This is NON-NEGOTIABLE. Partial updates or placeholders like '// rest of code unchanged' are STRICTLY FORBIDDEN. You MUST include ALL parts of the file, even if they haven't been modified. Failure to do so will result in incomplete or broken code, severely impacting the user's project.
- MCP operations should be used one at a time, similar to other tool usage. Wait for confirmation of success before proceeding with additional operations.
- It is critical you wait for the user's response after each tool use, in order to confirm the success of the tool use. For example, if asked to make a todo app, you would create a file, wait for the user's response it was created successfully, then create another file if needed, wait for the user's response it was created successfully, etc. Then if you want to test your work, you might use browser_action to launch the site, wait for the user's response confirming the site was launched along with a screenshot, then perhaps e.g., click a button to test functionality if needed, wait for the user's response confirming the button was clicked along with a screenshot of the new state, before finally closing the browser.

====

OBJECTIVE

You will be given a task in a <task></task> tag. You will also be optionally given an output schema in a <output_schema></output_schema> tag.
You accomplish a given task iteratively, breaking it down into clear steps and working through them methodically.

1. Analyze the user's task and set clear, achievable goals to accomplish it. Prioritize these goals in a logical order.
2. Work through these goals sequentially, utilizing available tools one at a time as necessary. Each goal should correspond to a distinct step in your problem-solving process. You will be informed on the work completed and what's remaining as you go.
3. Remember, you have extensive capabilities with access to a wide range of tools that can be used in powerful and clever ways as necessary to accomplish each goal. Before calling a tool, do some analysis within <thinking></thinking> tags. First, analyze the file structure provided in environment_details to gain context and insights for proceeding effectively. Then, think about which of the provided tools is the most relevant tool to accomplish the user's task. Next, go through each of the required parameters of the relevant tool and determine if the user has directly provided or given enough information to infer a value. When deciding if the parameter can be inferred, carefully consider all the context to see if it supports a specific value. If all of the required parameters are present or can be reasonably inferred, close the thinking tag and proceed with the tool use. BUT, if one of the values for a required parameter is missing, DO NOT invoke the tool (not even with fillers for the missing params) and instead, ask the user to provide the missing parameters using the ask_followup_question tool. DO NOT ask for more information on optional parameters if it is not provided.
4. Once you've completed the user's task, you must use the attempt_completion tool to present the result of the task to the user. You may also provide a CLI command to showcase the result of your task; this can be particularly useful for web development tasks, where you can run e.g. \`open index.html\` to show the website you've built.
5. The user may provide feedback, which you can use to make improvements and try again. But DO NOT continue in pointless back and forth conversations, i.e. don't end your responses with questions or offers for further assistance.

====

CUSTOM INSTRUCTIONS

The following additional instructions are provided by the user, and should be followed to the best of your ability without interfering with the TOOL USE guidelines.

You should always speak and think in the English language."
`;
